src/VisualStudio/Core/Def/Implementation/TableDataSource/Suppression/VisualStudioDiagnosticListSuppressionStateService.cs

```diff
@@ -0,0 +1,253 @@
+        {
+            var builder = ImmutableArray.CreateBuilder<DiagnosticData>();
+            var entries = selectedEntriesOnly ? _tableControl.SelectedEntries : _tableControl.Entries;
+            foreach (var entryHandle in entries)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                DiagnosticData diagnosticData = null;
+                int index;
+                var roslynSnapshot = GetEntriesSnapshot(entryHandle, out index);
+                if (roslynSnapshot != null)
+                {
+                    diagnosticData = roslynSnapshot.GetItem(index)?.Primary;
+                    if (diagnosticData != null && diagnosticData.HasTextSpan)
+                    {
+                        if (isAddSuppression)
+                        {
+                            // Compiler diagnostics can only be suppressed in source.
+                            if (!diagnosticData.IsSuppressed &&
+                                (isSuppressionInSource || !SuppressionHelpers.IsCompilerDiagnostic(diagnosticData)))
+                            {
+                                builder.Add(diagnosticData);
+                            }
+                        }
+                        else if (diagnosticData.IsSuppressed)
+                        {
+                            builder.Add(diagnosticData);
+                        }
+                    }
+                }
+            }
+
+            return builder.ToImmutable();
+        }
+
+        private static void UpdateSelectedItems(bool added, ref int count)
+        {
+            if (added)
+            {
+                count++;
+            }
+            else
+            {
+                count--;
+            }
+        }
+
+        private void HandleSuppressionStateEntry(bool isRoslynEntry, bool isSuppressedEntry, bool isCompilerDiagnosticEntry, bool added)
+        {
+            if (isRoslynEntry)
+            {
+                UpdateSelectedItems(added, ref _selectedRoslynItems);
+            }
+
+            if (isCompilerDiagnosticEntry)
+            {
+                UpdateSelectedItems(added, ref _selectedCompilerDiagnosticItems);
+            }
+
+            if (isSuppressedEntry)
+            {
+                UpdateSelectedItems(added, ref _selectedSuppressedItems);
+            }
+            else
+            {
+                UpdateSelectedItems(added, ref _selectedActiveItems);
+            }
+        }
+
+        private void HandleNonSuppressionStateEntry(bool added)
+        {
+            UpdateSelectedItems(added, ref _selectedNonSuppressionStateItems);
+        }
+
+        private void UpdateQueryStatus()
+        {
+            // Force the shell to refresh the QueryStatus for all the command since default behavior is it only does query
```

Comment:
I just took this code from the work you did for FXCop issues port to error list :) Basically, FxCop's event processor also does this when an item of interest was selected/deselected.